### PR TITLE
 tests/lib/fakestore/store: use graceful shutdown from net/http.Server

### PR DESCRIPTION
**Pull request by bboozzoo**
The race detector detected an issue with the 3rd party package we used:

```
==================
WARNING: DATA RACE
Read at 0x00c00024c7e0 by goroutine 484:
  net/http.(*conn).setState()
      /snap/go/10585/src/net/http/server.go:1840 +0x156
  net/http.(*conn).serve.func1()
      /snap/go/10585/src/net/http/server.go:1910 +0x374
  runtime.deferreturn()
      /snap/go/10585/src/runtime/panic.go:602 +0x5d
  net/http.(*Server).Serve.gowrap3()
      /snap/go/10585/src/net/http/server.go:3285 +0x4f

Previous write at 0x00c00024c7e0 by goroutine 475:
  gopkg.in/tylerb/graceful%2ev1.(*Server).manageConnections()
      /home/runner/work/snapd/snapd/src/github.com/snapcore/snapd/vendor/gopkg.in/tylerb/graceful.v1/graceful.go:399 +0x7d5
  gopkg.in/tylerb/graceful%2ev1.(*Server).Serve.gowrap1()
      /home/runner/work/snapd/snapd/src/github.com/snapcore/snapd/vendor/gopkg.in/tylerb/graceful.v1/graceful.go:297 +0x8f
```

Given that graceful shutdown has been in stdlib for a while now, replace the 3rd
part package with functionality provided by net/http.Server.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -29,7 +29,6 @@ require (
 	gopkg.in/macaroon.v1 v1.0.0-20150121114231-ab3940c6c165
 	gopkg.in/retry.v1 v1.0.3
 	gopkg.in/tomb.v2 v2.0.0-20161208151619-d5d1b5820637
-	gopkg.in/tylerb/graceful.v1 v1.2.15
```
Comment (zyga) [inline go.mod:32]: Woot, always nice to see fewer dependencies.

**Review by zyga — APPROVED** [review]
Nice, I remember when old Go had this issue, but it was fixed ages ago (by now). Do we have the same problem in the main daemon?

**Review by Meulengracht — APPROVED** [review]
Should we run the nested tests for safety?

**Comment by bboozzoo** [comment]
Added the nested tests label

File: go.sum [hunk 2]
```diff
@@ -74,7 +74,6 @@ golang.org/x/sys v0.7.0/go.mod h1:oPkhp1MJrh7nUepCBck5+mAzfO9JrbApNNgaTdGDITg=
 golang.org/x/term v0.7.0 h1:BEvjmm5fURWqcfbSKTdpkDXYBrUS1c0m8agp14W48vQ=
 golang.org/x/term v0.7.0/go.mod h1:P32HKFT3hSsZrRxla30E9HqToFYAQPCMs/zFMBUFqPY=
 golang.org/x/text v0.3.0/go.mod h1:NqM8EUOU14njkJ3fqMW+pc6Ldnwhi/IjpwHt7yyuwOQ=
-golang.org/x/text v0.9.0 h1:2sjJmO8cDvYveuX97RDLsxlyUxLl+GHoLxBiRdHllBE=
```
Comment (olivercalder) [inline go.sum:77]: Drive-by? Or is this a dependency of graceful?

File: tests/lib/fakestore/store/store.go [hunk 3]
```diff
@@ -134,12 +129,18 @@ func (s *Store) Start() error {
...
+	go func() {
+		closedC <- s.srv.Shutdown(ctx)
+	}()
+
+	err := <-closedC
+	if err != nil {
+		// forceful close
+		s.srv.Close()
+		return fmt.Errorf("store failed to stop after: %s", timeoutTime)
 	}
```
Comment (olivercalder) [inline tests/lib/fakestore/store/store.go:144]: Perhaps I'm missing something, but why is `closedC` necessary at all? Seems like we create the channel just to wait for `s.src.Shutdown(ctx)`. Why can't we call `s.src.Shutdown(ctx)` synchronously instead of `err := <-closedC`?

**Review by olivercalder — COMMENTED** [review]
Looks good! One question about the `Stop` function...

File: go.sum (hunk 2 above)
Comment (bboozzoo) [reply go.sum:77]: I think it was a dep of the graceful package. I only ran `go mod tidy`. Running it on master does not drop the module, so it must have been a transitive dependency.

File: tests/lib/fakestore/store/store.go (hunk 3 above)
Comment (bboozzoo) [reply tests/lib/fakestore/store/store.go:144]: Good point. I started with select like it was in the original code, but then added a context and forgot to drop to goroutine completely.

File: tests/lib/fakestore/store/store.go [hunk 4]
```diff
@@ -134,12 +129,12 @@ func (s *Store) Start() error {
...
 	timeoutTime := 2000 * time.Millisecond
-	s.srv.Stop(timeoutTime / 2)
-
-	select {
-	case <-s.srv.StopChan():
-	case <-time.After(timeoutTime):
-		return fmt.Errorf("store failed to stop after %s", timeoutTime)
+	ctx, cancel := context.WithTimeout(context.Background(), timeoutTime)
+	defer cancel()
+	if err := s.srv.Shutdown(ctx); err != nil {
```
Comment (olivercalder) [inline tests/lib/fakestore/store/store.go:134]: Nice, thanks!

**Review by olivercalder — APPROVED** [review]
Thanks for the fix, looks good!
